src/shared/services/path-util.service.ts

```diff
@@ -0,0 +1,70 @@
+/*
+ * This file is part of ng2-json-editor.
+ * Copyright (C) 2016 CERN.
+ *
+ * ng2-json-editor is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * ng2-json-editor is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with ng2-json-editor; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+ * In applying this license, CERN does not
+ * waive the privileges and immunities granted to it by virtue of its status
+ * as an Intergovernmental Organization or submit itself to any jurisdiction.
+ */
+
+import { Injectable } from '@angular/core';
+
+
+@Injectable()
+export class PathUtilService {
+
+  /**
+   *
+   * @param {Array<any>} path - Element's path
+   * @param {boolean} root - Find nearest or root parent array. True for root and false for nearest array parent
+   * @returns {Array<any>} - Returns the path to the array parent
+   */
+  getNearestOrRootArrayParentInPath(path: Array<any>, root: boolean): Array<any> {
```

Comment:
There are many cases in this PR where you use a flag in the method to separate the behaviour. You can read here <http://martinfowler.com/bliki/FlagArgument.html> an argument of why `flag` arguments are not so good.

While reading the calls to these methods, like `getNearestOrRootArrayParentInPath(path, true)` it's difficult to know what's going to happen.

So I would suggest refactoring them if possible into two different methods:

`getNearestArrayParentInPath` and `getRootArrayInPath` that just accept a `path` parameter. The common code in those functions you can extract into a helper function and leave inside of them just the differing code.

Apply this to all the cases where you have this pattern of using one function with a flag that changes the behaviour inside and it will improve a lot the readability.